text/0000-hierarchic-anonymous-life-time.md

```diff
@@ -0,0 +1,136 @@
+- Feature Name: Hierarchic anonymous life-time
+- Start Date: 2020-06-24
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+New use of anonymous life-time `'_` that implicitly added to current structure.
+
+# Motivation
+
+Motivation is to simplify iterative development and improving refactoring of the code
+
+Sometimes during refactoring such code:
+```rust
+struct CompositeObject {
+    obj: SomeType,
+}
+
+struct BigObject {
+    composite_obj: CompositeObject,
+    count: i32,
+}
+
+struct Application {
+   big_obj: BigObject,
+}
+```
+
+developer decides to make obj of SomeType as reference in CompositeObject type:
+```rust
+struct CompositeObject<'a> {
+    obj: &'a SomeType,
+}
+
+struct BigObject<'a> {
+    composite_obj: CompositeObject<'a>,
+    count: i32,
+}
+
+struct Application<'a> {
+   big_obj: BigObject<'a>,
+}
+```
+Everywhere in composition hierarchy I need to write 'a ... most of the times it is just boilerplate code ...
+
+What if instead of writing manually we will specify reference fields with anonymous life-time:
+```rust
+struct CompositeObject {
+    obj: &'_ SomeType,
+}
+
+struct BigObject {
+    composite_obj: CompositeObject,
+    count: i32,
+}
+
+struct Application {
+   big_obj: BigObject,
+}
+```
```

Comment:
I think that you need some notation on each point of use to specify that a lifetime was elided. That is, `CompositeObject` isn't a type with no lifetime parameters, it's a type whose exact lifetime parameters were deemed unimportant. So in `struct BigObject`, when defining a field with this type, we could write `CompositeObject<'_>` to mark this.

But even this wouldn't be sufficient: this feature wants to abstract away the number of lifetime parameters too (we could have two fields with elided lifetimes). So maybe `CompositeObject<'_ ...>` works? Where `'_ ...` means any number of lifetime parameters (that gets implicitly added to `BigObject`'s own parameters). Likewise for `Application`.

Now, you perhaps need some notation in the struct definitions too to specify that there are N elided lifetimes. For symmetry, `struct BigObject<'_ ...>` works.

With those additions, the feature seems less appealing. But without them, it's confusing.